core/src/impl/Kokkos_HostSharedPtr.hpp

```diff
@@ -0,0 +1,182 @@
+    other.m_control   = nullptr;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
+      const MaybeReferenceCountedPtr& other) noexcept
+      : m_value_ptr(other.m_value_ptr), m_control(other.m_control) {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    if (m_control) Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      MaybeReferenceCountedPtr&& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr       = other.m_value_ptr;
+      other.m_value_ptr = nullptr;
+      m_control         = other.m_control;
+      other.m_control   = nullptr;
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      const MaybeReferenceCountedPtr& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr = other.m_value_ptr;
+      m_control   = other.m_control;
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+      if (is_reference_counted())
+        Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION ~MaybeReferenceCountedPtr() { cleanup(); }
+
+  KOKKOS_FUNCTION T* get() const noexcept { return m_value_ptr; }
+  KOKKOS_FUNCTION T& operator*() const noexcept {
+    KOKKOS_EXPECTS(bool(*this));
+    return *get();
+  }
+  KOKKOS_FUNCTION T* operator->() const noexcept { return get(); }
+
+  // checks if the stored pointer is not null
+  KOKKOS_FUNCTION explicit operator bool() const noexcept {
+    return get() != nullptr;
+  }
+
+  // checks whether the MaybeReferenceCountedPtr manages the lifetime of the
+  // object
+  KOKKOS_FUNCTION bool is_reference_counted() const noexcept {
+    return m_control != nullptr;
+  }
+
+ private:
+  KOKKOS_FUNCTION void cleanup() noexcept {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    // If m_counter is set, then this instance is responsible for managing the
+    // objects pointed to by m_counter and m_value_ptr.
+    if (is_reference_counted()) {
+      KOKKOS_EXPECTS(m_control->m_deleter);
+      int const count = Kokkos::atomic_fetch_sub(&(m_control->m_counter), 1);
+      if (count == 1) {
+        (m_control->m_deleter)(m_value_ptr);
+        m_value_ptr = nullptr;
+        delete m_control;
+        m_control = nullptr;
+      }
+    }
+#endif
+  }
+
+  struct Control {
+    std::function<void(T*)> m_deleter;
+    int m_counter;
+  };
+
+  T* m_value_ptr;
+
+ protected:
+  Control* m_control;
+};
+
+template <class T>
+class HostSharedPtr : public MaybeReferenceCountedPtr<T> {
+ public:
+  explicit HostSharedPtr(T* value_ptr = nullptr)
+      : MaybeReferenceCountedPtr<T>(value_ptr, [](T* const t) { delete t; }) {}
+
+  template <class Deleter>
+  explicit HostSharedPtr(
+      T* value_ptr = nullptr, Deleter deleter = [](T* const t) { delete t; })
+      : MaybeReferenceCountedPtr<T>(value_ptr, std::move(deleter)) {}
+
+  int use_count() const noexcept {
```

Comment:
Should this really return `0` in the case where it is non-owning?  I'm not sure what it should return in that case (sorry I don't have a better answer).

Reply:
If we don't have a better answer, we might as well just keep using 0. :slightly_smiling_face: